Gcp artifact (#235)
jenkins-x/terraform-google-jx: README.md

You are prompted for any further required configuration.
The number of prompts depends on how much you have [pre-configured](#inputs) via your Terraform variables.

### Artifact Registry in Multicluster Configuration

> In a multiclustser environment, it is crucial to set `artifact_enable = true` **only in one cluster**—preferably the development cluster. This is important because enabling the Artifact Registry in multiple clusters may cause Terraform to fail, as it cannot create multiple instances of the same repository.

Reviewer: "set `artifact_enable = true`" is a confusing since that is the default. How about "leave the value of `artifact_enable` as `true`"

#### Configuration Note
The jx-requirements.yml will be automatically updated by the Jenkins X bootjob to remove the registry entry. If for any reason this update does not occur as expected, you should manually update the file by creating a branch and submitting a pull request (PR) to remove the registry field.

#### Migration Options
Here are two strategies for transitioning container images from `gcr.io` to the Artifact Registry `us-central1-docker.pkg.dev/my-project/oci`:

**Option 1: Do Not Migrate Existing Images**
- Continue developing applications as usual. New images, upon their release, will be pushed to the Artifact Registry.
- **Important**: Ensure that all builds are triggered and applications are promoted before Google completely shuts down the Container Registry. This step is critical to avoid disruptions in service.
To identify which images from `gcr.io` are currently used in your cluster, you can use the following command line:
```bash
kubectl get pods --all-namespaces -o jsonpath="{range .items[*].spec['initContainers', 'containers'][*]}{.image}{'\n'}{end}" | fgrep gcr.io/project_id | sort -u
```

**Option 2: Migrate Existing Images**

> To migrate existing images from `gcr.io` to the Artifact Registry `us-central1-docker.pkg.dev/my-project/oci`, follow the steps in the [Google documentation](https://cloud.google.com/artifact-registry/

Reviewer: It looks like you mean that existing images should be migrated to a new repo created by this module. That will not help since the images specified in helm charts still point to gcr.io.  That is why I referred you to https://cloud.google.com/artifact-registry/docs/transition/setup-gcr-repo .

You are prompted for any further required configuration.
The number of prompts depends on how much you have [pre-configured](#inputs) via your Terraform variables.

> ### Enable Artifact Registry for Docker Images

Reviewer: ```suggestion
### Migration from Container to Artifact Registry
```

> Google has deprecated `gcr.io` and now recommends using the Google Artifact Registry. By default, this service can be enabled with the variable `artifact_enable = true`.

Reviewer: I find this wording a bit hard to understand. Maybe something like this instead.
```suggestion
Google has deprecated `gcr.io` and now recommends the use of Artifact Registry. The default of this module is now to create and use a repository in Artifact Registry for container images.
```

> #### Multicluster Configuration:

Reviewer: You typically don't use colon in headlines (only in running text)
```suggestion
#### Multicluster Configuration
```

In a multiclustser environment, it is crucial to set `artifact_enable = true` **only in one cluster**—preferably the development cluster. This is important because enabling the Artifact Registry in multiple clusters may cause Terraform to fail, as it cannot create multiple instances of the same repository.

- **Development Cluster:** Set `artifact_enable = true` to utilize the Artifact Registry.
- **Other Clusters:** Ensure `artifact_enable = false` to prevent Terraform errors due to existing repository conflicts.

Google GKE clusters automatically have permissions to download from the Artifact Registry. For multiclustser setups across different projects, additional permission configurations may be necessary.

#### Configuration Note:
When transitioning to use the Artifact Registry, you must update your configuration in the `Cluster git repository`:
- **Remove the `registry` entry** from `jx-requirements.yaml` by creating a branch and submitting a pull request (PR).
- After the PR is merged, Jenkins will automatically repopulate the `registry` line with the value provided by Terraform.

> #### Migration Options:

Reviewer: ```suggestion
#### Migration Options
```

> There are two strategies for transitioning Docker images from `gcr.io` to the Artifact Registry:

Reviewer: Well there are more possible strategies than that. Also since the general term used in Google's documentation is *container* images I think it would be clearer to use that.
```suggestion
Here are two strategies for transitioning container images from `gcr.io` to the Artifact Registry:
```

**Option 1: Do Not Migrate Existing Images**
- Continue developing applications as usual. New images, upon their release, will be pushed to the Artifact Registry.

> - Trigger builds for existing applications to generate new images in the Artifact Registry.

Reviewer: I think you should clarify that these builds needs to be triggered *and* the applications need to be promoted *before* Google shuts down container registry.

It would also be helpful to give a command line for which images from ones container registry are still used in a cluster. Something like this:

`kubectl get pods --all-namespaces -o jsonpath="{range .items[*].spec['initContainers', 'containers'][*]}{.image}{'\n'}{end}" | fgrep gcr.io/project_id | sort -u`

**Option 2: Migrate Existing Images**

> To migrate existing images from `gcr.io` to the Artifact Registry, follow the steps in the [Google documentation](https://cloud.google.com/artifact-registry/

Reviewer: You should describe what values should be specified for this terraform module in this case.

**Option 1: Do Not Migrate Existing Images**
- Continue developing applications as usual. New images, upon their release, will be pushed to the Artifact Registry.
- **Important**: Ensure that all builds are triggered and applications are promoted before Google completely shuts down the Container Registry. This step is critical to avoid disruptions in service.
To identify which images from `gcr.io` are currently used in your cluster, you can use the following command line (replace `project_id` with your actual `gcp project-id`):
```bash
kubectl get pods --all-namespaces -o jsonpath="{range .items[*].spec['initContainers', 'containers'][*]}{.image}{'\n'}{end}" | fgrep gcr.io/project_id | sort -u
```
The Artifact Registry url is `<reistry-location>-docker.pkg.dev/my_project/oci`
where <reistry-location> is the location of your Artifact Registry es: `us-central1` , project_is is your `gcp Project ID` 
`oci` is the name of artifact repository defined with terraform variable `artifact_repository_id` default value `oci` 
es: `us-central1-docker.pkg.dev/my-project/oci`.

**Option 2: Migrate Existing Images**

> To migrate existing images from `gcr.io` to the Artifact Registry `us-central1-docker.pkg.dev/my-project/oci`, you should follow the steps in the [Google documentation](https://cloud.google.com/artifact-registry/docs/transition/setup-gcr-repo)

Author: Because in my case i didn't, and i not sure what is the step.

> To migrate existing images from `gcr.io` to the Artifact Registry `us-central1-docker.pkg.dev/my-project/oci`, you should follow the steps in the [Google documentation](https://cloud.google.com/artifact-registry/docs/transition/setup-gcr-repo)

Author: Also it will not works for Multi cluster if they are in different Project
```The following limitations apply to repositories with gcr.io domain support:

You cannot map a Container Registry host to an Artifact Registry repository in a different project.```
This was the reason I not follow the google Guide and i not made any migration.

> To migrate existing images from `gcr.io` to the Artifact Registry `us-central1-docker.pkg.dev/my-project/oci`, you should follow the steps in the [Google documentation](https://cloud.google.com/artifact-registry/docs/transition/setup-gcr-repo)

Author: I follow the all steps in the Google Docs.
